From 79faea3aeed0a871b119567bd32a9b9c033f41b0 Mon Sep 17 00:00:00 2001
From: Wu Zhenjiang <wuzhenjiang@loongson.cn>
Date: Thu, 13 Jun 2024 09:31:28 +0800
Subject: [PATCH 11/15] FROMLIST: LoongArch64: Optimize functions in
 intrapred8.S file - Patch 2

The optimized functions are as follows:
intra_filter_4x4_lsx/intra_filter_4x4_lasx
intra_filter_8x8_lsx/intra_filter_8x8_lasx
intra_filter_16x16_lsx/intra_filter_16x16_lasx
intra_filter_32x32_lsx/intra_filter_32x32_lasx
intra_pred_planar_4x4_lsx
intra_pred_planar_8x8_lsx
intra_pred_planar_16x16_lsx
intra_pred_planar_32x32_lsx

Link: https://bitbucket.org/multicoreware/x265_git/pull-requests/31
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 source/common/loongarch64/asm-primitives.cpp |  16 +
 source/common/loongarch64/intrapred.h        |  16 +
 source/common/loongarch64/intrapred8.S       | 719 +++++++++++++++++++
 3 files changed, 751 insertions(+)

diff --git a/source/common/loongarch64/asm-primitives.cpp b/source/common/loongarch64/asm-primitives.cpp
index 883c63b60..08ad496f1 100644
--- a/source/common/loongarch64/asm-primitives.cpp
+++ b/source/common/loongarch64/asm-primitives.cpp
@@ -352,6 +352,16 @@ void setupAssemblyPrimitives(EncoderPrimitives &p, int cpuMask)
         CHROMA_PU_422(8,  64, lsx);
 
         // intra_pred
+        p.cu[BLOCK_4x4].intra_filter = PFX(intra_filter_4x4_lsx);
+        p.cu[BLOCK_8x8].intra_filter = PFX(intra_filter_8x8_lsx);
+        p.cu[BLOCK_16x16].intra_filter = PFX(intra_filter_16x16_lsx);
+        p.cu[BLOCK_32x32].intra_filter = PFX(intra_filter_32x32_lsx);
+
+        p.cu[BLOCK_4x4].intra_pred[PLANAR_IDX]   = PFX(intra_pred_planar_4x4_lsx);
+        p.cu[BLOCK_8x8].intra_pred[PLANAR_IDX]   = PFX(intra_pred_planar_8x8_lsx);
+        p.cu[BLOCK_16x16].intra_pred[PLANAR_IDX] = PFX(intra_pred_planar_16x16_lsx);
+        p.cu[BLOCK_32x32].intra_pred[PLANAR_IDX] = PFX(intra_pred_planar_32x32_lsx);
+
         p.cu[BLOCK_4x4].intra_pred[2] = PFX(intra_pred_ang4_2_lsx);
         p.cu[BLOCK_4x4].intra_pred[3] = PFX(intra_pred_ang4_3_lsx);
         p.cu[BLOCK_4x4].intra_pred[4] = PFX(intra_pred_ang4_4_lsx);
@@ -612,6 +622,12 @@ void setupAssemblyPrimitives(EncoderPrimitives &p, int cpuMask)
         ASSIGN2(p.chroma[X265_CSP_I422].pu[CHROMA_422_32x48].addAvg, addAvg_32x48_lasx);
         ASSIGN2(p.chroma[X265_CSP_I422].pu[CHROMA_422_32x64].addAvg, addAvg_32x64_lasx);
 
+        // intra_filter
+        p.cu[BLOCK_4x4].intra_filter = PFX(intra_filter_4x4_lasx);
+        p.cu[BLOCK_8x8].intra_filter = PFX(intra_filter_8x8_lasx);
+        p.cu[BLOCK_16x16].intra_filter = PFX(intra_filter_16x16_lasx);
+        p.cu[BLOCK_32x32].intra_filter = PFX(intra_filter_32x32_lasx);
+
         //intra_pred
         p.cu[BLOCK_8x8].intra_pred[3] = PFX(intra_pred_ang8_3_lasx);
         p.cu[BLOCK_8x8].intra_pred[4] = PFX(intra_pred_ang8_4_lasx);
diff --git a/source/common/loongarch64/intrapred.h b/source/common/loongarch64/intrapred.h
index 5220e2166..f0835eb21 100644
--- a/source/common/loongarch64/intrapred.h
+++ b/source/common/loongarch64/intrapred.h
@@ -30,6 +30,22 @@
 extern "C" {
 #endif /* __cplusplus */
 
+void x265_intra_filter_4x4_lsx(const pixel *samples, pixel *filtered);
+void x265_intra_filter_8x8_lsx(const pixel *samples, pixel *filtered);
+void x265_intra_filter_16x16_lsx(const pixel *samples, pixel *filtered);
+void x265_intra_filter_32x32_lsx(const pixel *samples, pixel *filtered);
+void x265_intra_filter_4x4_lasx(const pixel *samples, pixel *filtered);
+void x265_intra_filter_8x8_lasx(const pixel *samples, pixel *filtered);
+void x265_intra_filter_16x16_lasx(const pixel *samples, pixel *filtered);
+void x265_intra_filter_32x32_lasx(const pixel *samples, pixel *filtered);
+
+void x265_intra_pred_planar_4x4_lsx(pixel *dst, intptr_t dstStride, const pixel *srcPix, int /*dirMode*/, int /*bFilter*/);
+void x265_intra_pred_planar_8x8_lsx(pixel *dst, intptr_t dstStride, const pixel *srcPix, int /*dirMode*/, int /*bFilter*/);
+void x265_intra_pred_planar_16x16_lsx(pixel *dst, intptr_t dstStride, const pixel *srcPix, int /*dirMode*/, int /*bFilter*/);
+void x265_intra_pred_planar_32x32_lsx(pixel *dst, intptr_t dstStride, const pixel *srcPix, int /*dirMode*/, int /*bFilter*/);
+void x265_intra_pred_planar_16x16_lasx(pixel *dst, intptr_t dstStride, const pixel *srcPix, int /*dirMode*/, int /*bFilter*/);
+void x265_intra_pred_planar_32x32_lasx(pixel *dst, intptr_t dstStride, const pixel *srcPix, int /*dirMode*/, int /*bFilter*/);
+
 void x265_intra_pred_ang8_2_lsx(pixel* dst, intptr_t dstStride, const pixel* srcPix, int dirMode, int bFilter);
 void x265_intra_pred_ang8_3_lsx(pixel* dst, intptr_t dstStride, const pixel* srcPix, int dirMode, int bFilter);
 void x265_intra_pred_ang8_4_lsx(pixel* dst, intptr_t dstStride, const pixel* srcPix, int dirMode, int bFilter);
diff --git a/source/common/loongarch64/intrapred8.S b/source/common/loongarch64/intrapred8.S
index 19969ca8a..8c9e53c7b 100644
--- a/source/common/loongarch64/intrapred8.S
+++ b/source/common/loongarch64/intrapred8.S
@@ -23,6 +23,725 @@
 
 #include "loongson_asm.S"
 
+.macro INTRAFILTER_LSX_16 offset, src0, src1, src2, src3, src4, src5, dst0, dst1
+    // load samples[offset ~ offset+15], samples[offset+1 ~ offset+16], samples[offset+2 ~ offset+17]
+    vld             \src0,  a0,     \offset
+    vld             \src2,  a0,     \offset + 1
+    vld             \src3,  a0,     \offset + 9
+    vld             \src4,  a0,     \offset + 2
+    vexth.hu.bu     \src1,  \src0
+    vsllwil.hu.bu   \src0,  \src0,  0
+    vsllwil.hu.bu   \src2,  \src2,  1
+    vsllwil.hu.bu   \src3,  \src3,  1
+    vexth.hu.bu     \src5,  \src4
+    vsllwil.hu.bu   \src4,  \src4,  0
+
+    // calculate filtered[offset+1 ~ offset+16]
+    vadd.h          \dst0,  \src0,  \src2
+    vadd.h          \dst1,  \src1,  \src3
+    vadd.h          \dst0,  \dst0,  \src4
+    vadd.h          \dst1,  \dst1,  \src5
+    vaddi.hu        \dst0,  \dst0,  2
+    vaddi.hu        \dst1,  \dst1,  2
+.endm
+
+// void x265_intra_filter_4x4_lsx(const pixel* samples, pixel* filtered)
+function x265_intra_filter_4x4_lsx
+    ld.bu           t0,     a0,     0
+    ld.bu           t1,     a0,     8
+    ld.bu           t2,     a0,     16
+
+    // calculate filtered[1~16]
+    INTRAFILTER_LSX_16      0,  vr0,  vr1,  vr2,  vr3,  vr4,  vr5,  vr6,  vr7
+
+    // calculate filtered[0]
+    vpickve2gr.hu   t5,     vr0,    1
+    alsl.d          t5,     t0,     t5,     1
+    vpickve2gr.hu   t6,     vr1,    1
+    add.d           t5,     t6,     t5
+    addi.d          t5,     t5,     2
+    srli.d          t5,     t5,     2
+    st.b            t5,     a1,     0
+
+    // calculate filtered[tuSize2 + 1]
+    alsl.d          t6,     t6,     t0,     1
+    vpickve2gr.hu   t4,     vr1,    2
+    add.d           t6,     t6,     t4
+    addi.d          t6,     t6,     2
+    vinsgr2vr.h     vr7,    t6,     0
+    vsrlni.b.h      vr7,    vr6,    2
+
+    // calculate filtered[tuSize2] and filtered[tuSize2 + tuSize2]
+    vinsgr2vr.b     vr7,    t1,     7
+    vinsgr2vr.b     vr7,    t2,     15
+    vst             vr7,    a1,     1       // store filtered[1~16]
+endfunc
+
+// void x265_intra_filter_8x8_lsx(const pixel* samples, pixel* filtered)
+function x265_intra_filter_8x8_lsx
+    ld.bu           t0,     a0,     0
+    ld.bu           t1,     a0,     16
+    ld.bu           t2,     a0,     32
+
+    // calculate filtered[1~32]
+    INTRAFILTER_LSX_16      0,  vr0,  vr1,  vr4,  vr5,  vr8,  vr9,  vr12, vr13
+    INTRAFILTER_LSX_16      16, vr2,  vr3,  vr6,  vr7,  vr10, vr11, vr14, vr15
+
+    // calculate filtered[0]
+    vpickve2gr.hu   t5,     vr0,    1
+    alsl.d          t5,     t0,     t5,     1
+    vpickve2gr.hu   t6,     vr2,    1
+    add.d           t5,     t6,     t5
+    addi.d          t5,     t5,     2
+    srli.d          t5,     t5,     2
+    st.b            t5,     a1,     0
+
+    // calculate filtered[tuSize2 + 1]
+    alsl.d          t6,     t6,     t0,     1
+    vpickve2gr.hu   t4,     vr2,    2
+    add.d           t6,     t6,     t4
+    addi.d          t6,     t6,     2
+    vinsgr2vr.h     vr14,   t6,     0
+    vsrlni.b.h      vr13,   vr12,   2
+    vsrlni.b.h      vr15,   vr14,   2
+
+    // calculate filtered[tuSize2] and filtered[tuSize2 + tuSize2]
+    vinsgr2vr.b     vr13,   t1,     15
+    vinsgr2vr.b     vr15,   t2,     15
+    vst             vr13,   a1,     1
+    vst             vr15,   a1,     17
+endfunc
+
+// void x265_intra_filter_16x16_lsx(const pixel* samples, pixel* filtered)
+function x265_intra_filter_16x16_lsx
+    ld.bu           t0,     a0,     0
+    ld.bu           t1,     a0,     32
+    ld.bu           t2,     a0,     64
+
+    // calculate filtered[1~32]
+    INTRAFILTER_LSX_16      0,  vr0,  vr1,  vr4,  vr5,  vr8,  vr9,  vr12, vr13
+    INTRAFILTER_LSX_16      16, vr2,  vr3,  vr6,  vr7,  vr10, vr11, vr14, vr15
+
+    // calculate filtered[0]
+    vpickve2gr.hu   t5,     vr0,    1
+    alsl.d          t5,     t0,     t5,     1
+    vpickve2gr.hu   t6,     vr11,   7
+    add.d           t5,     t6,     t5
+    addi.d          t5,     t5,     2
+    srli.d          t5,     t5,     2
+    st.b            t5,     a1,     0
+    vsrlni.b.h      vr13,   vr12,   2
+    vsrlni.b.h      vr15,   vr14,   2
+    vinsgr2vr.b     vr15,   t1,     15
+    vst             vr13,   a1,     1
+    vst             vr15,   a1,     17
+
+    // calculate filtered[33~64]
+    INTRAFILTER_LSX_16      32, vr0,  vr1,  vr4,  vr5,  vr8,  vr9,  vr12, vr13
+    INTRAFILTER_LSX_16      48, vr2,  vr3,  vr6,  vr7,  vr10, vr11, vr14, vr15
+
+    // calculate filtered[tuSize2 + 1]
+    alsl.d          t6,     t6,     t0,     1
+    vpickve2gr.hu   t4,     vr0,    2
+    add.d           t6,     t6,     t4
+    addi.d          t6,     t6,     2
+    vinsgr2vr.h     vr12,   t6,     0
+    vsrlni.b.h      vr13,   vr12,   2
+    vsrlni.b.h      vr15,   vr14,   2
+    vinsgr2vr.b     vr15,   t2,     15
+    vst             vr13,   a1,     33
+    vst             vr15,   a1,     49
+endfunc
+
+// void x265_intra_filter_32x32_lsx(const pixel* samples, pixel* filtered)
+function x265_intra_filter_32x32_lsx
+    ld.bu           t0,     a0,     0
+    ld.bu           t1,     a0,     64
+    ld.bu           t2,     a0,     128
+
+    // calculate filtered[1~32]
+    INTRAFILTER_LSX_16      0,  vr0,  vr1,  vr4,  vr5,  vr8,  vr9,  vr12, vr13
+    INTRAFILTER_LSX_16      16, vr2,  vr3,  vr6,  vr7,  vr10, vr11, vr14, vr15
+
+    vsrlni.b.h      vr13,   vr12,   2
+    vsrlni.b.h      vr15,   vr14,   2
+    vst             vr13,   a1,     1
+    vst             vr15,   a1,     17
+    vpickve2gr.hu   t5,     vr0,    1       // save samples[1] for calculating samples[65]
+
+    // calculate filtered[33~64]
+    INTRAFILTER_LSX_16      32, vr0,  vr1,  vr4,  vr5,  vr8,  vr9,  vr12, vr13
+    INTRAFILTER_LSX_16      48, vr2,  vr3,  vr6,  vr7,  vr10, vr11, vr14, vr15
+
+    // calculate filtered[0]
+    alsl.d          t5,     t0,     t5,     1
+    vpickve2gr.hu   t6,     vr11,   7
+    add.d           t5,     t6,     t5
+    addi.d          t5,     t5,     2
+    srli.d          t5,     t5,     2
+    st.b            t5,     a1,     0
+    vsrlni.b.h      vr13,   vr12,   2
+    vsrlni.b.h      vr15,   vr14,   2
+    vinsgr2vr.b     vr15,   t1,     15
+    vst             vr13,   a1,     33
+    vst             vr15,   a1,     49
+
+    // calculate filtered[65~96]
+    INTRAFILTER_LSX_16      64, vr0,  vr1,  vr4,  vr5,  vr8,  vr9,  vr12, vr13
+    INTRAFILTER_LSX_16      80, vr2,  vr3,  vr6,  vr7,  vr10, vr11, vr14, vr15
+
+    // calculate filtered[tuSize2 + 1]
+    alsl.d          t6,     t6,     t0,     1
+    vpickve2gr.hu   t4,     vr0,    2
+    add.d           t6,     t6,     t4
+    addi.d          t6,     t6,     2
+    vinsgr2vr.h     vr12,   t6,     0
+    vsrlni.b.h      vr13,   vr12,   2
+    vsrlni.b.h      vr15,   vr14,   2
+    vst             vr13,   a1,     65
+    vst             vr15,   a1,     81
+
+    // calculate filtered[97~128]
+    INTRAFILTER_LSX_16      96, vr0,  vr1,  vr4,  vr5,  vr8,  vr9,  vr12, vr13
+    INTRAFILTER_LSX_16      112,vr2,  vr3,  vr6,  vr7,  vr10, vr11, vr14, vr15
+    vsrlni.b.h      vr13,   vr12,   2
+    vsrlni.b.h      vr15,   vr14,   2
+    vinsgr2vr.b     vr15,   t2,     15
+    vst             vr13,   a1,     97
+    vst             vr15,   a1,     113
+endfunc
+
+.macro INTRAFILTER_LASX_16 src0, src1, src2, dst0, offset
+    // load samples[offset ~ offset+15], samples[offset+1 ~ offset+16], samples[offset+2 ~ offset+17]
+    xvld            \src0,  a0,     \offset
+    xvld            \src1,  a0,     \offset + 1
+    xvld            \src2,  a0,     \offset + 2
+    vext2xv.hu.bu   \src0,  \src0
+    vext2xv.hu.bu   \src1,  \src1
+    xvslli.h        \src1,  \src1,  1
+    vext2xv.hu.bu   \src2,  \src2
+
+    // calculate filtered[[offset+1 ~ offset+16]
+    xvadd.h         \dst0,  \src0,  \src1
+    xvadd.h         \dst0,  \dst0,  \src2
+    xvssrarni.bu.h  \dst0,  \dst0,  2
+.endm
+
+// void x265_intra_filter_4x4_lasx(const pixel* samples, pixel* filtered)
+function x265_intra_filter_4x4_lasx
+    ld.bu           t0,     a0,     0
+    ld.bu           t1,     a0,     8
+    ld.bu           t2,     a0,     16
+
+    // calculate filtered[1~16]
+    INTRAFILTER_LASX_16     xr0,  xr1,  xr2,  xr3,  0
+
+    // calculate filtered[0]
+    ld.bu           t5,     a0,     1
+    alsl.d          t5,     t0,     t5,     1
+    ld.bu           t6,     a0,     9
+    add.d           t5,     t6,     t5
+    addi.d          t5,     t5,     2
+    srli.d          t5,     t5,     2
+    st.b            t5,     a1,     0
+
+    // store filtered[1~16]
+    xvpermi.d       xr3,    xr3,    0b00001000
+    vst             vr3,    a1,     1
+    st.b            t1,     a1,     8               // store filtered[8]
+
+    // calculate filtered[tuSize2 + 1]
+    slli.d          t6,     t6,     1
+    add.d           t6,     t6,     t0
+    ld.bu           t4,     a0,     10
+    add.d           t6,     t6,     t4
+    addi.d          t6,     t6,     2
+    srli.d          t6,     t6,     2
+    st.b            t6,     a1,     9
+    st.b            t2,     a1,     16              // store filtered[16]
+endfunc
+
+// void x265_intra_filter_8x8_lasx(const pixel* samples, pixel* filtered)
+function x265_intra_filter_8x8_lasx
+    ld.bu           t0,     a0,     0
+    ld.bu           t1,     a0,     16
+    ld.bu           t2,     a0,     32
+
+    // calculate filtered[1 ~ 32]
+    INTRAFILTER_LASX_16     xr0,  xr2,  xr4,  xr6,  0
+    INTRAFILTER_LASX_16     xr1,  xr3,  xr5,  xr7,  16
+
+    // calculate filtered[0]
+    ld.bu           t5,     a0,     1
+    alsl.d          t5,     t0,     t5,     1
+    ld.bu           t6,     a0,     17
+    add.d           t5,     t6,     t5
+    addi.d          t5,     t5,     2
+    srli.d          t5,     t5,     2
+    st.b            t5,     a1,     0
+
+    // store filtered[1~32]
+    xvshuf4i.d      xr6,    xr7,    0b00001000
+    xvpermi.d       xr6,    xr6,    0b11011000
+    xvst            xr6,    a1,     1
+    st.b            t1,     a1,     16              // store filtered[16]
+
+    slli.d          t6,     t6,     1
+    add.d           t6,     t6,     t0
+    ld.bu           t4,     a0,     18
+    add.d           t6,     t6,     t4
+    addi.d          t6,     t6,     2
+    srli.d          t6,     t6,     2
+    st.b            t6,     a1,     17              // store filtered[17]
+    st.b            t2,     a1,     32              // store filtered[32]
+endfunc
+
+// void x265_intra_filter_16x16_lasx(const pixel* samples, pixel* filtered)
+function x265_intra_filter_16x16_lasx
+    ld.bu           t0,     a0,     0
+    ld.bu           t1,     a0,     32
+    ld.bu           t2,     a0,     64
+
+    // calculate filtered[1 ~ 64]
+    INTRAFILTER_LASX_16     xr0,  xr4,  xr8,  xr12, 0
+    INTRAFILTER_LASX_16     xr1,  xr5,  xr9,  xr13, 16
+    INTRAFILTER_LASX_16     xr2,  xr6,  xr10, xr14, 32
+    INTRAFILTER_LASX_16     xr3,  xr7,  xr11, xr15, 48
+
+    // calculate filtered[0]
+    ld.bu           t5,     a0,     1
+    alsl.d          t5,     t0,     t5,     1
+    ld.bu           t6,     a0,     33
+    add.d           t5,     t6,     t5
+    addi.d          t5,     t5,     2
+    srli.d          t5,     t5,     2
+    st.b            t5,     a1,     0
+
+    // store filtered[1~32]
+    xvshuf4i.d      xr12,   xr13,   0b00001000
+    xvpermi.d       xr12,   xr12,   0b11011000
+    xvst            xr12,   a1,     1
+    st.b            t1,     a1,     32              // store filtered[32]
+
+    // store filtered[33~64]
+    xvshuf4i.d      xr14,   xr15,   0b00001000
+    xvpermi.d       xr14,   xr14,   0b11011000
+    xvst            xr14,   a1,     33
+    alsl.d          t6,     t6,     t0,     1
+    ld.bu           t4,     a0,     34
+    add.d           t6,     t6,     t4
+    addi.d          t6,     t6,     2
+    srli.d          t6,     t6,     2
+    st.b            t6,     a1,     33              // store filtered[33]
+    st.b            t2,     a1,     64              // store filtered[64]
+endfunc
+
+// void x265_intra_filter_32x32_lasx(const pixel* samples, pixel* filtered)
+function x265_intra_filter_32x32_lasx
+    ld.bu           t0,     a0,     0
+    ld.bu           t1,     a0,     64
+    ld.bu           t2,     a0,     128
+
+    //calculate filtered[1~64]
+    INTRAFILTER_LASX_16     xr0,  xr4,  xr8,  xr12, 0
+    INTRAFILTER_LASX_16     xr1,  xr5,  xr9,  xr13, 16
+    INTRAFILTER_LASX_16     xr2,  xr6,  xr10, xr14, 32
+    INTRAFILTER_LASX_16     xr3,  xr7,  xr11, xr15, 48
+
+    // store filtered[1~32]
+    xvshuf4i.d      xr12,   xr13,   0b00001000
+    xvpermi.d       xr12,   xr12,   0b11011000
+    xvst            xr12,   a1,     1
+    // store filtered[33~64]
+    xvshuf4i.d      xr14,   xr15,   0b00001000
+    xvpermi.d       xr14,   xr14,   0b11011000
+    xvst            xr14,   a1,     33
+
+    // calculate filtered[65~128]
+    INTRAFILTER_LASX_16     xr0,  xr4,  xr8,  xr12, 64
+    INTRAFILTER_LASX_16     xr1,  xr5,  xr9,  xr13, 80
+    INTRAFILTER_LASX_16     xr2,  xr6,  xr10, xr14, 96
+    INTRAFILTER_LASX_16     xr3,  xr7,  xr11, xr15, 112
+
+    // calculate filtered[0]
+    ld.bu           t5,     a0,     1
+    alsl.d          t5,     t0,     t5,     1
+    ld.bu           t6,     a0,     65
+    add.d           t5,     t6,     t5
+    addi.d          t5,     t5,     2
+    srli.d          t5,     t5,     2
+    st.b            t5,     a1,     0
+
+    // store filtered[65~96]
+    xvshuf4i.d      xr12,   xr13,   0b00001000
+    xvpermi.d       xr12,   xr12,   0b11011000
+    xvst            xr12,   a1,     65
+    // store filtered[97~128]
+    xvshuf4i.d      xr14,   xr15,   0b00001000
+    xvpermi.d       xr14,   xr14,   0b11011000
+    xvst            xr14,   a1,     97
+    st.b            t1,     a1,     64              // store filtered[64]
+    alsl.d          t6,     t6,     t0,     1
+    ld.bu           t4,     a0,     66
+    add.d           t6,     t6,     t4
+    addi.d          t6,     t6,     2
+    srli.d          t6,     t6,     2
+    st.b            t6,     a1,     65              // store filtered[65]
+    st.b            t2,     a1,     128             // store filtered[128]
+endfunc
+
+.macro CALCULATE_INTRA_PRED_PLANAR  type, src0, src1, src2, src3, \
+                                    src4, src5, src6, dst0, dst1
+    \type\()vreplgr2vr.h    \dst0,    zero
+    \type\()vreplgr2vr.h    \dst1,    zero
+    \type\()vmadd.h         \dst0,    \src0,    \src1
+    \type\()vmadd.h         \dst1,    \src2,    \src3
+    \type\()vadd.h          \dst0,    \dst0,    \src4
+    \type\()vadd.h          \dst1,    \dst1,    \src5
+    \type\()vadd.h          \dst0,    \dst0,    \src6
+    \type\()vadd.h          \dst0,    \dst1,    \dst0
+.endm
+
+// void x265_intra_pred_planar_4x4_lsx(pixel* dst, intptr_t dstStride, const pixel* srcPix, int /*dirMode*/, int /*bFilter*/)
+const planar_4x4_lsx
+.short 3, 2, 1, 0, 1, 2, 3, 4
+endconst
+function x265_intra_pred_planar_4x4_lsx     // blkSize = 1 << 2 = 4
+    // prepare data
+    ld.bu           t2,     a2,     5
+    ld.bu           t3,     a2,     13
+    fld.d           f0,     a2,     1
+    fld.d           f1,     a2,     9
+    vsllwil.hu.bu   vr0,    vr0,    0
+    vsllwil.hu.bu   vr1,    vr1,    0
+
+    la.local        t4,     planar_4x4_lsx
+    vld             vr2,    t4,     0
+    vld             vr3,    t4,     8
+
+    vreplgr2vr.h    vr4,    t2
+    vreplgr2vr.h    vr5,    t3
+    vreplgr2vr.h    vr6,    t3
+    vmul.h          vr4,    vr3,    vr4
+    addi.d          t7,     zero,   4
+    vreplgr2vr.h    vr7,    t7
+
+    addi.d          t4,     zero,   0
+    addi.d          t5,     zero,   3
+    addi.d          t6,     a0,     0
+
+.LOOP_INTRA_PRED_PLANAR_4x4_LSX:
+    vreplve.h       vr9,    vr1,    t4
+    vreplgr2vr.h    vr10,    t5
+    CALCULATE_INTRA_PRED_PLANAR  , vr2, vr9, vr10, vr0, vr4, vr6, vr7, vr8, vr11
+    vsrlni.b.h      vr8,    vr8,    3
+    vstelm.w        vr8,    t6,     0,      0       // store dst[y*dstStride+x]
+
+    // prepare for next loop
+    addi.d          t4,     t4,     1
+    addi.d          t5,     t5,     -1
+    vadd.h          vr6,    vr6,    vr5
+    add.d           t6,     t6,     a1
+    blt             t4,     t7,     .LOOP_INTRA_PRED_PLANAR_4x4_LSX
+endfunc
+
+// void x265_intra_pred_planar_8x8_lsx(pixel* dst, intptr_t dstStride, const pixel* srcPix, int /*dirMode*/, int /*bFilter*/)
+const planar_8x8_lsx
+.short 7, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 6, 7, 8
+endconst
+function x265_intra_pred_planar_8x8_lsx     // blkSize = 1 << 3 = 8
+    // prepare data
+    ld.bu           t2,     a2,     9
+    ld.bu           t3,     a2,     25
+    fld.d           f0,     a2,     1
+    fld.d           f1,     a2,     17
+    vsllwil.hu.bu   vr0,    vr0,    0
+    vsllwil.hu.bu   vr1,    vr1,    0
+
+    la.local        t4,     planar_8x8_lsx
+    vld             vr2,    t4,     0
+    vld             vr3,    t4,     16
+
+    vreplgr2vr.h    vr4,    t2
+    vreplgr2vr.h    vr5,    t3
+    vreplgr2vr.h    vr6,    t3
+    vmul.h          vr4,    vr3,    vr4
+    addi.d          t7,     zero,   8
+    vreplgr2vr.h    vr7,    t7
+
+    addi.d          t4,     zero,   0
+    addi.d          t5,     zero,   7
+    addi.d          t6,     a0,     0
+
+.LOOP_INTRA_PRED_PLANAR_8x8_LSX:
+    vreplve.h       vr9,    vr1,    t4
+    vreplgr2vr.h    vr10,    t5
+    CALCULATE_INTRA_PRED_PLANAR  , vr2, vr9, vr10, vr0, vr4, vr6, vr7, vr8, vr11
+    vsrlni.b.h      vr8,    vr8,    4
+    vstelm.d        vr8,    t6,     0,      0       // store dst[y*dstStride+x]
+
+    // prepare for next loop
+    addi.d          t4,     t4,     1
+    addi.d          t5,     t5,     -1
+    vadd.h          vr6,    vr6,    vr5
+    add.d           t6,     t6,     a1
+    blt             t4,     t7,     .LOOP_INTRA_PRED_PLANAR_8x8_LSX
+endfunc
+
+// void x265_intra_pred_planar_16x16_lsx(pixel* dst, intptr_t dstStride, const pixel* srcPix, int /*dirMode*/, int /*bFilter*/)
+const planar_16x16_lsx
+.short 15, 14, 13, 12, 11, 10, 9,  8
+.short 7,  6,  5,  4,  3,  2,  1,  0
+.short 1,  2,  3,  4,  5,  6,  7,  8
+.short 9,  10, 11, 12, 13, 14, 15, 16
+endconst
+function x265_intra_pred_planar_16x16_lsx   // blkSize = 1 << 4 = 16
+    // prepare data
+    ld.bu           t2,     a2,     17
+    ld.bu           t3,     a2,     49
+    fld.d           f0,     a2,     1
+    fld.d           f10,    a2,     9
+    vld             vr1,    a2,     33
+    vsllwil.hu.bu   vr0,    vr0,    0
+    vsllwil.hu.bu   vr10,   vr10,   0
+
+    la.local        t4,     planar_16x16_lsx
+    vld             vr2,    t4,     0
+    vld             vr12,   t4,     16
+    vld             vr3,    t4,     32
+    vld             vr13,   t4,     48
+
+    vreplgr2vr.h    vr4,    t2
+    vreplgr2vr.h    vr5,    t3
+    vreplgr2vr.h    vr6,    t3
+    vmul.h          vr14,   vr13,   vr4
+    vmul.h          vr4,    vr3,    vr4
+    addi.d          t7,     zero,   16
+    vreplgr2vr.h    vr7,    t7
+
+    addi.d          t4,     zero,   0
+    addi.d          t5,     zero,   15
+    addi.d          t6,     a0,     0
+
+.LOOP_INTRA_PRED_PLANAR_16x16_LSX:
+    vreplve.b       vr9,    vr1,    t4
+    vsllwil.hu.bu   vr9,    vr9,    0
+    vreplgr2vr.h    vr11,   t5
+    CALCULATE_INTRA_PRED_PLANAR  , vr2,  vr9, vr11, vr0,  vr4,  vr6, vr7, vr8,  vr15
+    CALCULATE_INTRA_PRED_PLANAR  , vr12, vr9, vr11, vr10, vr14, vr6, vr7, vr18, vr16
+    vsrlni.b.h      vr18,   vr8,    5
+    vst             vr18,   t6,     0       // store dst[y*dstStride+x]
+
+    // prepare for next loop
+    addi.d          t4,     t4,     1
+    addi.d          t5,     t5,     -1
+    vadd.h          vr6,    vr6,    vr5
+    add.d           t6,     t6,     a1
+    blt             t4,     t7,     .LOOP_INTRA_PRED_PLANAR_16x16_LSX
+endfunc
+
+// void x265_intra_pred_planar_32x32_lsx(pixel* dst, intptr_t dstStride, const pixel* srcPix, int /*dirMode*/, int /*bFilter*/)
+const planar_32x32_lsx
+.short 31, 30, 29, 28, 27, 26, 25, 24
+.short 23, 22, 21, 20, 19, 18, 17, 16
+.short 15, 14, 13, 12, 11, 10,  9,  8
+.short  7,  6,  5,  4,  3,  2,  1,  0
+.short  1,  2,  3,  4,  5,  6,  7,  8
+.short  9, 10, 11, 12, 13, 14, 15, 16
+.short 17, 18, 19, 20, 21, 22, 23, 24
+.short 25, 26, 27, 28, 29, 30, 31, 32
+endconst
+function x265_intra_pred_planar_32x32_lsx   // blkSize = 1 << 5 = 32
+    // prepare data
+    ld.bu           t2,     a2,     33
+    ld.bu           t3,     a2,     97
+    fld.d           f0,     a2,     1
+    fld.d           f1,     a2,     9
+    fld.d           f2,     a2,     17
+    fld.d           f3,     a2,     25
+    vld             vr4,    a2,     65
+    vld             vr21,   a2,     81
+    vsllwil.hu.bu   vr0,    vr0,    0
+    vsllwil.hu.bu   vr1,    vr1,    0
+    vsllwil.hu.bu   vr2,    vr2,    0
+    vsllwil.hu.bu   vr3,    vr3,    0
+
+    la.local        t4,     planar_32x32_lsx
+    vld             vr5,    t4,     0
+    vld             vr6,    t4,     16
+    vld             vr7,    t4,     32
+    vld             vr8,    t4,     48
+    vld             vr9,    t4,     64
+    vld             vr10,   t4,     80
+    vld             vr11,   t4,     96
+    vld             vr12,   t4,     112
+
+    vreplgr2vr.h    vr13,   t2
+    vmul.h          vr9,    vr13,   vr9
+    vmul.h          vr10,   vr13,   vr10
+    vmul.h          vr11,   vr13,   vr11
+    vmul.h          vr12,   vr13,   vr12
+    vreplgr2vr.h    vr13,   t3
+    vreplgr2vr.h    vr14,   t3
+    addi.d          t7,     zero,   32
+    vreplgr2vr.h    vr15,   t7
+
+    addi.d          t4,     zero,   0
+    addi.d          t5,     zero,   31
+    addi.d          t6,     a0,     0
+    addi.d          t8,     zero,   16
+    vreplve.b       vr20,   vr4,    t4
+.LOOP_INTRA_PRED_PLANAR_32x32_LSX:
+    vsllwil.hu.bu   vr20,   vr20,   0
+    vreplgr2vr.h    vr22,   t5
+    CALCULATE_INTRA_PRED_PLANAR  , vr5,  vr20, vr22, vr0,  vr9,  vr14, vr15, vr16, vr23
+    CALCULATE_INTRA_PRED_PLANAR  , vr6,  vr20, vr22, vr1,  vr10, vr14, vr15, vr17, vr23
+    vsrlni.b.h      vr17,   vr16,   6
+    vst             vr17,   t6,     0
+    CALCULATE_INTRA_PRED_PLANAR  , vr7,  vr20, vr22, vr2,  vr11, vr14, vr15, vr18, vr23
+    CALCULATE_INTRA_PRED_PLANAR  , vr8,  vr20, vr22, vr3,  vr12, vr14, vr15, vr19, vr23
+    vsrlni.b.h      vr19,   vr18,   6
+    vst             vr19,   t6,     16      // store dst[y*dstStride+(16~31)]
+
+    // prepare for next loop
+    addi.d          t4,     t4,     1
+    addi.d          t5,     t5,     -1
+    vadd.h          vr14,   vr14,   vr13
+    add.d           t6,     t6,     a1
+    // if y >= 16, switch to high-short part loop
+    bge             t4,     t8,     .INTRA_PRED_PLANAR_32x32_LSX_SWITCH_TO_HIHG_HALF_PART_LOOP
+    vreplve.b       vr20,   vr4,    t4
+
+.INTRA_PRED_PLANAR_32x32_LSX_WHILE_JUDGE:
+    blt             t4,     t7,     .LOOP_INTRA_PRED_PLANAR_32x32_LSX
+    b               .INTRA_PRED_PLANAR_32x32_LSX_TERMINATE
+
+.INTRA_PRED_PLANAR_32x32_LSX_SWITCH_TO_HIHG_HALF_PART_LOOP:
+    // new_index = y - 16, for get the high-short part data of left[]
+    sub.d           a3,     t4,     t8
+    vreplve.b       vr20,   vr21,   a3      // rep left[y] to vr20, y = {16 ~ 31}
+    b               .INTRA_PRED_PLANAR_32x32_LSX_WHILE_JUDGE
+.INTRA_PRED_PLANAR_32x32_LSX_TERMINATE:
+endfunc
+
+// void x265_intra_pred_planar_16x16_lasx(pixel* dst, intptr_t dstStride, const pixel* srcPix, int /*dirMode*/, int /*bFilter*/)
+const planar_16x16_lasx
+.short 15, 14, 13, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0
+.short  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16
+endconst
+function x265_intra_pred_planar_16x16_lasx  // blkSize = 1 << 4 = 16
+    // prepare data
+    ld.bu           t2,     a2,     17
+    ld.bu           t3,     a2,     49
+    vld             vr0,    a2,     1
+    vld             vr1,    a2,     33
+    vext2xv.hu.bu   xr0,    xr0
+
+    la.local        t4,     planar_16x16_lasx
+    xvld            xr2,    t4,     0
+    xvld            xr3,    t4,     32
+
+    xvreplgr2vr.h   xr4,    t2
+    xvreplgr2vr.h   xr5,    t3
+    xvreplgr2vr.h   xr6,    t3
+    xvmul.h         xr4,    xr3,    xr4
+    addi.d          t7,     zero,   16
+    xvreplgr2vr.h   xr7,    t7
+
+    addi.d          t4,     zero,   0
+    addi.d          t5,     zero,   15
+    addi.d          t6,     a0,     0
+
+.LOOP_INTRA_PRED_PLANAR_16x16_LASX:
+    xvreplve.b      xr9,    xr1,    t4
+    vext2xv.hu.bu   xr9,    xr9
+    xvreplgr2vr.h   xr10,   t5
+    CALCULATE_INTRA_PRED_PLANAR       x, xr2, xr9, xr10, xr0, xr4, xr6, xr7, xr8, xr11
+    xvsrlni.b.h     xr8,    xr8,    5
+    xvpermi.d       xr8,    xr8,    0b00001000
+    vst             vr8,    t6,     0
+
+    // prepare for next loop
+    addi.d          t4,     t4,     1
+    addi.d          t5,     t5,     -1
+    xvadd.h         xr6,    xr6,    xr5
+    add.d           t6,     t6,     a1
+    blt             t4,     t7,     .LOOP_INTRA_PRED_PLANAR_16x16_LASX
+endfunc
+
+// void x265_intra_pred_planar_32x32_lasx(pixel* dst, intptr_t dstStride, const pixel* srcPix, int /*dirMode*/, int /*bFilter*/)
+const planar_32x32_lasx
+.short 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16
+.short 15, 14, 13, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0
+.short  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16
+.short 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32
+endconst
+function x265_intra_pred_planar_32x32_lasx   // blkSize = 1 << 5 = 32
+    // prepare data
+    ld.bu           t2,     a2,     33
+    ld.bu           t3,     a2,     97
+    vld             vr0,    a2,     1
+    vld             vr10,   a2,     17
+    vld             vr1,    a2,     65
+    vld             vr11,   a2,     81
+    vext2xv.hu.bu   xr0,    xr0
+    vext2xv.hu.bu   xr10,   xr10
+
+    la.local        t4,     planar_32x32_lasx
+    xvld            xr2,    t4,     0
+    xvld            xr12,   t4,     32
+    xvld            xr3,    t4,     64
+    xvld            xr13,   t4,     96
+
+    xvreplgr2vr.h   xr4,    t2
+    xvreplgr2vr.h   xr5,    t3
+    xvreplgr2vr.h   xr6,    t3
+    xvmul.h         xr14,   xr13,   xr4
+    xvmul.h         xr4,    xr3,    xr4
+    addi.d          t7,     zero,   32
+    xvreplgr2vr.h   xr7,    t7
+
+    addi.d          t4,     zero,   0
+    addi.d          t5,     zero,   31
+    addi.d          t6,     a0,     0
+    addi.d          a4,     zero,   16
+    xvreplve.b      xr9,    xr1,    t4
+    vext2xv.hu.bu   xr9,    xr9
+.LOOP_INTRA_PRED_PLANAR_32x32_LASX:
+    xvreplgr2vr.h   xr19,   t5
+    CALCULATE_INTRA_PRED_PLANAR     x, xr2,  xr9, xr19, xr0,  xr4,  xr6,  xr7, xr8,  xr13
+    CALCULATE_INTRA_PRED_PLANAR     x, xr12, xr9, xr19, xr10, xr14, xr6,  xr7, xr18, xr23
+    xvsrlni.b.h     xr8,    xr8,    6
+    xvsrlni.b.h     xr18,   xr18,   6
+    xvpermi.d       xr8,    xr8,    0b00001000
+    xvpermi.d       xr18,   xr18,   0b00001000
+    xvpermi.q       xr18,   xr8,    0b00100000
+    xvst            xr18,   t6,     0
+
+    // prepare for next loop
+    addi.d          t4,     t4,     1
+    addi.d          t5,     t5,     -1
+    xvadd.h         xr6,    xr6,    xr5
+    add.d           t6,     t6,     a1
+    // if y >= 16, switch to high-short part loop
+    bge             t4,     a4,     .INTRA_PRED_PLANAR_32x32_LASX_SWITCH_TO_HIHG_HALF_PART_LOOP
+    xvreplve.b      xr9,    xr1,    t4
+    vext2xv.hu.bu   xr9,    xr9
+
+.INTRA_PRED_PLANAR_32x32_LASX_WHILE_JUDGE:
+    blt             t4,     t7,     .LOOP_INTRA_PRED_PLANAR_32x32_LASX
+    b               .INTRA_PRED_PLANAR_32x32_LASX_TERMINATE
+
+.INTRA_PRED_PLANAR_32x32_LASX_SWITCH_TO_HIHG_HALF_PART_LOOP:
+    // new_index = y - 16, for get the high-short part data of left[]
+    sub.d           a3,     t4,     a4
+    xvreplve.b      xr9,    xr11,   a3
+    vext2xv.hu.bu   xr9,    xr9
+    b               .INTRA_PRED_PLANAR_32x32_LASX_WHILE_JUDGE
+
+.INTRA_PRED_PLANAR_32x32_LASX_TERMINATE:
+endfunc
+
 /*
 void intra_pred_ang8_c(pixel* dst, intptr_t dstStride, const pixel *srcPix0,
                       int dirMode, int bFilter)
-- 
2.52.0

